database.go

```diff
@@ -0,0 +1,723 @@
+		// series. So don't do the sort if we don't have to.
+		if len(ids) > 1 && ids[len(ids)-1] < ids[len(ids)-2] {
+			sort.Sort(ids)
+		}
+		valueMap[v] = ids
+	}
+
+	return true
+}
+
+// seriesByTags returns the Series that matches the given tagset.
+func (m *Measurement) seriesByTags(tags map[string]string) *Series {
+	return m.series[string(marshalTags(tags))]
+}
+
+// sereisIDs returns the series ids for a given filter
+func (m *Measurement) seriesIDs(filter *Filter) (ids SeriesIDs) {
+	values := m.seriesByTagKeyValue[filter.Key]
+	if values == nil {
+		return
+	}
+
+	// hanlde regex filters
+	if filter.Regex != nil {
+		for k, v := range values {
+			if filter.Regex.MatchString(k) {
+				if ids == nil {
+					ids = v
+				} else {
+					ids = ids.Union(v)
+				}
+			}
+		}
+		if filter.Not {
+			ids = m.ids.Reject(ids)
+		}
+		return
+	}
+
+	// this is for the value is not null query
+	if filter.Not && filter.Value == "" {
+		for _, v := range values {
+			if ids == nil {
+				ids = v
+			} else {
+				ids.Intersect(v)
+			}
+		}
+		return
+	}
+
+	// get the ids that have the given key/value tag pair
+	ids = SeriesIDs(values[filter.Value])
+
+	// filter out these ids from the entire set if it's a not query
+	if filter.Not {
+		ids = m.ids.Reject(ids)
+	}
+
+	return
+}
+
+// tagValues returns an array of unique tag values for the given key
+func (m *Measurement) tagValues(key string) TagValues {
+	tags := m.seriesByTagKeyValue[key]
+	values := make(map[string]bool, len(tags))
+	for k, _ := range tags {
+		values[k] = true
+	}
+	return TagValues(values)
+}
+
+type Measurements []*Measurement
+
+// Field represents a series field.
+type Field struct {
+	ID   uint8     `json:"id,omitempty"`
+	Name string    `json:"name,omitempty"`
+	Type FieldType `json:"field"`
+}
+
+type FieldType int
+
+const (
+	Int64 FieldType = iota
+	Float64
+	String
+	Boolean
+	Binary
+)
+
+// Fields represents a list of fields.
+type Fields []*Field
+
+// Series belong to a Measurement and represent unique time series in a database
+type Series struct {
+	ID   uint32
+	Tags map[string]string
+}
+
+// RetentionPolicy represents a policy for creating new shards in a database and how long they're kept around for.
+type RetentionPolicy struct {
+	// Unique name within database. Required.
+	Name string
+
+	// Length of time to keep data around
+	Duration time.Duration
+
+	ReplicaN uint32
+	SplitN   uint32
+
+	Shards []*Shard
+}
+
+// NewRetentionPolicy returns a new instance of RetentionPolicy with defaults set.
+func NewRetentionPolicy(name string) *RetentionPolicy {
+	return &RetentionPolicy{
+		Name:     name,
+		ReplicaN: DefaultReplicaN,
+		SplitN:   DefaultSplitN,
+		Duration: DefaultShardRetention,
+	}
+}
+
+// shardByTimestamp returns the shard in the space that owns a given timestamp for a given series id.
+// Returns nil if the shard does not exist.
+func (rp *RetentionPolicy) shardByTimestamp(id uint32, timestamp time.Time) *Shard {
```

Comment:
As I mentioned above, this function would be more self-documenting if `id` was replaced by `seriesId`. This is just a style question, of course.